NSFS | Versioning | Implement GET/HEAD object

vh05 opened the pull request:
### Explain the changes
This feature is part of NSFS versioning.
GET/HEAD object will return the latest or the requested version object.

Below is the behaviour of the patch.
NOTE: Lets call "file under bucket_path as regular file in this commit context"

1. with version disabled
   - no behavior change - should return the regular file

2. with version enabled/suspended
   - file created before the version is enabled - should return the regular file
   - no version id specified - should return the regular file
   - with version specified - if version id of regular file matches with version specified - should return the regular file
   - with version specified - if version id of regular file does not match with version specified - should return the versioned file
   - delete marked xattr placed - should return ENOENT

### Issues: Fixed #7128

### Testing Instructions:
1. 


- [ ] Doc added/updated
- [ ] Tests added

### TBD
Optimize calls to `stat` in` _find_version_path()` 

`Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

v-harihar commented:
/cc @romayalon

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -514,7 +517,36 @@ class NamespaceFS {
             const stat = await nb_native().fs.stat(fs_context, file_path);
             if (isDirectory(stat)) throw Object.assign(new Error('NoSuchKey'), { code: 'ENOENT' });
-            return this._get_object_info(params.bucket, params.key, stat);
+
```
Hi @v-harihar A great start! :)
this functionality of searching a version in .versions/ or fallback to the parent folder is needed for almost every API function.
In order to write and maintain this logic once, it should move out to a helper function outside of read_object_md.
I would create a function with similar functionality to what you did but I would leave the delete marker error handling to read_object_md/read_object_stream.
that way, instead of calling: `const file_path = this._get_file_path(params);` I would call a function that does this searching logic.
After that change, you will need to do the delete marker check logic once

v-harihar replied:
sure .. I will rebase this patch once your patch is merged.

v-harihar commented:
/retest

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -545,6 +554,11 @@ class NamespaceFS {
         const fs_context = this.prepare_fs_context(object_sdk);
         let file_path;
+
+        if (this.versioning === 'ENABLED' || this.versioning === 'SUSPENDED') {
```
this is a bit hacky, I think it's preferred to have here the same behavior as in read_object_md:
get the version path and handle delete marker and version id etc

v-harihar replied:
Since we already have object md, I felt we can use the whatever information we got from md call which does save some IO operations. However I have replaced `get_file_path()` with `get_version_path()`, May be we can think of optimization here to avoid multiple IO operations

romayalon replied:
How did it even pass test_namespace_fs.js tests? 
test_namespace_fs calls directly to the namespace runtime object and it's not going through the s3_put_object.js flow.
Optimization is always good and I guess we can optimize the 2 calls to get_version_path() to 1 but we need to take into consideration the scale and the fact that the file can be changed between the read_object_stream and read_object_md

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -509,12 +512,18 @@ class NamespaceFS {
             const stat = await nb_native().fs.stat(fs_context, file_path);
             if (isDirectory(stat)) throw Object.assign(new Error('NoSuchKey'), { code: 'ENOENT' });
-            return this._get_object_info(params.bucket, params.key, stat);
+            if (this.versioning === 'ENABLED' || this.versioning === 'SUSPENDED') {
```
please define the following consts and use them instead of comparing the versioning mode with a string:
const VER_ENABLED = 'ENABLED';
const VER_SUSPENDED = 'SUSPENDED';
const VER_DISABLED = 'DISABLED';

v-harihar replied:
sure

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -509,12 +512,18 @@ class NamespaceFS {
-            return this._get_object_info(params.bucket, params.key, stat);
+            if (this.versioning === 'ENABLED' || this.versioning === 'SUSPENDED') {
+                const xattr_delete_marker = stat.xattr[XATTR_DELETE_MARKER];
+                if (xattr_delete_marker) {
```
you can move this check to a  helper function named for example
_throw_if_delete_marker() since we will need it in read_object_stream

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -1332,6 +1346,29 @@ class NamespaceFS {
+            const stat = await nb_native().fs.stat(fs_context, file_path);
+            if (isDirectory(stat)) throw Object.assign(new Error('NoSuchKey'), { code: 'ENOENT' });
+            const xattr_version_id = stat.xattr[XATTR_VERSION_ID];
+            if (!xattr_version_id || !params.version_id || xattr_version_id === params.version_id) {
```
A few comments about that if - 
1. no need to stat if !params.version_id, you can move this check to the condition in the first line, and the versioning mode check is not needed because the user can ask for a version id even if the versioning is still disabled and we should fail since the version doesn't exist, but we shouldn't return the latest version. 
2. The first condition here !xattr_version_id is incorrect, for example, if a user asked for a specific version but it doesn't exist and there is only latest version of that object, the latest version will return but it's not what the user asked, we should fail with no such object in this case.
3. The last condition - we should fail with no such object here also if we can't find the version in the latest version path or under .versions/ so we need to add stat check for the version path as well and fail if we couldn't find this version

v-harihar replied:
I updated the patch to align with above comments (may not be complete, Please take look)

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -1332,6 +1346,29 @@ class NamespaceFS {
+            } else {
+                // If the object wanted is among the previous versions, return the
+                // object from $BUCKET_PATH/.versions/version_id
+                const v_path = path.normalize(path.join(this.bucket_path, VERSION_DIR, params.version_id));
```
the version path should be of the following form:
```
_get_version_path(key, version_id) {
        const key_version = path.basename(key) + (version_id ? '_' + version_id : '');
        return path.normalize(path.join(this.bucket_path, path.dirname(key), hidden_versions_path, key_version));
    }
```

v-harihar replied:
Considered `key_version` to address the file inside `.versions`. Please correct me if wrong

romayalon commented on src/test/unit_tests/test_namespace_fs.js:
```diff
@@ -229,6 +234,283 @@ mocha.describe('namespace_fs', function() {
+            }
+        });
+
+        mocha.it('get object, versioning not enabled - should return latest object', async function() {
```
could you add a test for get/head null version when it's not the latest? 
currently, it's not a different flow than what we have but maybe for future changes it's good to have it as well

romayalon commented on src/test/unit_tests/test_namespace_fs.js:
```diff
@@ -764,3 +1046,14 @@ async function read_object_stream(ns, bucket, file_key, dummy_object_sdk) {
+        await target_file.replacexattr(DEFAULT_FS_CONFIG, xattr_version_id);
+    }
+
+    return target_file;
```
no newline at the end of the file here and in test_bucketspace_versioning as well

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -509,12 +514,13 @@ class NamespaceFS {
             if (isDirectory(stat)) throw Object.assign(new Error('NoSuchKey'), { code: 'ENOENT' });
-            return this._get_object_info(params.bucket, params.key, stat);
+            this._throw_if_delete_marker(stat);
+            return this._get_object_info(params.bucket, params.key, stat, params.version_id || 'null');
```
why do we need this 'null'? 
Also, when bucket versioning is disabled we shouldn't return version_id

v-harihar replied:
No it is the past behavior. `_get_object_info()` used to return `version_id` as `null`

romayalon replied:
you are right we used to always return '1'? so the sdk/cli removes it when getting an answer?

v-harihar replied:
let me check this

v-harihar commented on src/sdk/namespace_fs.js:
```diff
@@ -547,7 +553,7 @@ class NamespaceFS {
             await this._load_bucket(params, fs_context);
-            file_path = this._get_file_path(params);
+            file_path = await this._get_version_path(params, fs_context);
             await this._check_path_in_bucket_boundaries(fs_context, file_path);
```
sure

romayalon commented on src/sdk/namespace_fs.js:
```diff
@@ -1332,6 +1341,37 @@ class NamespaceFS {
+        const file_path = this._get_file_path(params);
+        if (params.version_id) {
+            const stat = await nb_native().fs.stat(fs_context, file_path);
+            if (isDirectory(stat)) throw Object.assign(new Error('NoSuchKey'), { code: 'ENOENT' });
```
I ran some examples on AWS and noticed that head object with version-id='1' throws BadRequest and not ENOENT (as the current implementation will throw), so we need here a check for an invalid version parameter.
It should be a helper function so it can be used in all the functions

romayalon commented:
@v-harihar Looks good, one last comment from my side

vh05 committed:
NSFS | Versioning | Implement GET/HEAD object

This feature is part of NSFS versioning.
GET/HEAD object will return the latest or the requested version object.

Below is the behaviour of the patch.
NOTE: Lets call "file under bucket_path as regular file in this commit context"

1. with version disabled
   no behavior change - should return the regular file
   with version enabled/suspended

2. file created before the version is enabled - should return the regular file
   no version id specified - should return the regular file
   with version specified - if version id of regular file matches with version specified - should return the regular file
   with version specified - if version id of regular file does not match with version specified - should return the versioned file
   delete marked xattr placed - should return ENOENT

Signed-off-by: Vinayakswami Hariharmath <vharihar@redhat.com>

liranmauda reviewed (changes requested):
@v-harihar Please squash the commits
We are not using merge from the UI, as we do not want to get the `Merge branch 'master' into <branch>` commit
